### PR TITLE
[TP+FSDP] Integration code

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82166

FDSP TP integration. Not aimed for landing.

Differential Revision: [D38086237](https://our.internmc.facebook.com/intern/diff/D38086237/)

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_tp_integration.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+        # Generate sub-process group for TP and FSDP.
+        tp_pg, fsdp_pg = self._generate_sub_pg(model_parallel_size)
+
+        # Wrap the control group model with FSDP only.
+        model = FSDP(model, process_group=self.process_group, cpu_offload=cpu_offload)
+        sharding_specs = _generate_chunk_sharding_spec(tp_pg.size(), tp_pg)
+        # Shard the module first and then wrap with FSDP.
+        sharding_plan = _module_sharding_plan(sharding_specs)
+        shard_module(model_tp, sharding_plan, process_group=tp_pg)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_tp_integration.py:230]: When rebased on `viable/strict`, if I print `model_tp.named_parameters()` after the `shard_module()` call and before wrapping with FSDP, I only see
```
net1.bias torch.Size([8]) <class 'torch.nn.parameter.Parameter'>
net2.bias torch.Size([4]) <class 'torch.nn.parameter.Parameter'>
net3.weight torch.Size([12, 4]) <class 'torch.nn.parameter.Parameter'>
net3.bias torch.Size([12]) <class 'torch.nn.parameter.Parameter'>
```
on rank 0. It seems that `net1.weight` and `net2.weight` are missing, which I believe to be the intended `ShardedTensor`s (based on `_module_sharding_plan()`).

It seems that they are not being made visible to `nn.Module` methods like `named_parameters()`. This leads to an error like:
```
Traceback (most recent call last):
  File "/fsx/users/andgu/work/pytorch/torch/testing/_internal/common_distributed.py", line 622, in run_test
    getattr(self, test_name)()
  File "/fsx/users/andgu/work/pytorch/torch/testing/_internal/common_distributed.py", line 503, in wrapper
    fn()
  File "/fsx/users/andgu/work/pytorch/torch/testing/_internal/common_utils.py", line 216, in instantiated_test
    test(self, **param_kwargs)
  File "/fsx/users/andgu/work/pytorch/torch/testing/_internal/common_distributed.py", line 145, in wrapper
    return func(*args, **kwargs)
  File "/fsx/users/andgu/work/pytorch/test/distributed/fsdp/test_fsdp_tp_integration.py", line 251, in test_fsdp_tp_integration
    self._sync_tp_module_grads(
  File "/fsx/users/andgu/work/pytorch/test/distributed/fsdp/test_fsdp_tp_integration.py", line 179, in _sync_tp_module_grads
    all_params = torch.cat(all_params).contiguous().split(splits)
  File "/fsx/users/andgu/work/pytorch/torch/_tensor.py", line 775, in split
    return super(Tensor, self).split_with_sizes(split_size, dim)
RuntimeError: start (48) + length (48) exceeds dimension size (72).
```
where `72 = 8 + 4 + 12*4 + 12`.

cc: @wanchaol

File: test/distributed/fsdp/test_fsdp_tp_integration.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+OPS_NOT_SHARD = [
+    "net3.weight",
+    "net3.bias",
+]
+
+SHARD_PARAMS = [
+    "net1.weight",
+    "net2.weight",
+]
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_tp_integration.py:50]: Curious: does it matter that not all parameters in `SimpleModel` appear in either of these two lists?

**Review by awgu — COMMENTED** [review]
I did an initial dive into the code.

I think there are two behaviors that may be bugs:
1. `nn.Parameter`s that wrap `ShardedTensor` not being returned in `named_parameters()`, and
2. `isinstance(sharded_tensor, FlatParameter)` returning `True`.

A third point to mention is `ShardedTensor` data may be "double-counted" in `FlatParameter`s in this implementation. (Due to 1, this may not be happening at the moment, but if that is fixed, then this may happen.)

For these three points, please refer to the inline comments for more information.

cc: @zhaojuanmao @rohan-varma

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 3]
```diff
@@ -217,8 +232,8 @@ def get_param_views(
                 f"Incorrect numel of supplied data: got {data.numel()} but expected {self.full_numel}"
             )
         return (
-            t.view(s)
-            for (t, s) in zip(data.split(self._param_numels), self._param_shapes)
+            t.view(s) if not info else self._rebuild_sharded_tensor(t.view(s), info)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/fsdp/flatten_params_wrapper.py:235]: Yep this is also my understanding.

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 4]
```diff
@@ -352,6 +369,13 @@ def _init_flatten_params(
         params = []
         for module_name, m in self.named_modules():
             for n, p in m.named_parameters(recurse=False):
+                if p is not None and isinstance(p, ShardedTensor):
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/fsdp/flatten_params_wrapper.py:372]: Yes.

File: test/distributed/fsdp/test_fsdp_tp_integration.py (hunk 1 above)
Comment (fduwjj) [reply test/distributed/fsdp/test_fsdp_tp_integration.py:230]: This is indeed wired. Let me ask Wanchao about this.

File: test/distributed/fsdp/test_fsdp_tp_integration.py (hunk 2 above)
Comment (fduwjj) [reply test/distributed/fsdp/test_fsdp_tp_integration.py:50]: It should be fine. This is just for grad sync. For real use case, we don't have something like this.

File: torch/distributed/fsdp/flatten_params_wrapper.py [hunk 5]
```diff
@@ -137,6 +138,14 @@ def __new__(
             cls, data, requires_grad=requires_grad
         )  # type: ignore[call-arg]
 
+    def _rebuild_sharded_tensor(self, p, sharding_info):
+        return ShardedTensor._init_from_local_tensor(
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/fsdp/flatten_params_wrapper.py:142]: Yeah I think so. It's like adding a wrapper.

**Comment by fduwjj** [comment]
OK, Addressed the concerns offline with @awgu  and now looks like we got all concerns addressed. Here is a rebased version which also includes distributed tensor: https://github.com/pytorch/pytorch/pull/82581. I am now closing this PR.

**Pull request closed without merging**